### PR TITLE
Swift 4

### PR DESCRIPTION
**Pull request by chadcummings**
Swift 4 Upgrades! Let me know if I missed anything.

- Upgrades to project and code to allow for Swift 4, SDK iOS11.
- Target iOS 9.3.
- Add support for UPC-A.
- Add all new barcode types to the metadata array.

### CODE REVIEW

File: Sources/BarcodeScannerController.swift [hunk 1]
```diff
@@ -343,8 +343,6 @@ open class BarcodeScannerController: UIViewController {
         center(subview: focusView, inSize: CGSize(width: 218, height: 150))
     }
     center(subview: settingsButton, inSize: CGSize(width: 150, height: 50))
-
-    headerView.isHidden = !isBeingPresented
```
Comment (vadymmarkov) [inline Sources/BarcodeScannerController.swift]: @chadcummings What is the reason behind this change?

File: BarcodeScanner.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -228,7 +234,7 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 9.2;
+				IPHONEOS_DEPLOYMENT_TARGET = 9.3;
```
Comment (vadymmarkov) [inline BarcodeScanner.xcodeproj/project.pbxproj]: Could it be `9.0` or there are some specific APIs you want to use? @chadcummings

Comment (chadcummings) [reply BarcodeScanner.xcodeproj/project.pbxproj]: After reviewing which devices can run iOS 9.0, I discovered that all of the target devices can also run iOS 9.3. This led me to the conclusion that is was pointless to support iOS 9.0 when all devices that run iOS 9.0 will run iOS 9.3. Supporting iOS 9.0 has no benefit, as no additional devices are gained. That's my only reasoning, really.

Ultimately, this code is your's. I am happy to use whichever version you would like, but I would prefer trimming extra baggage as much as possible.

File: Sources/BarcodeScannerController.swift (hunk 1 above)
Comment (chadcummings) [reply Sources/BarcodeScannerController.swift]: This fixes a bug when the screen orientation was changed, the Header would disappear and never reappear, even when the device was rotated back. I haven't noticed any negative side effects, but I'll look to you for guidance here.

File: BarcodeScanner.xcodeproj/project.pbxproj (hunk 2 above)
Comment (vadymmarkov) [reply BarcodeScanner.xcodeproj/project.pbxproj]: I totally agree with you, but I think we still have projects with minimum target set to 9.0, which is probably something we might want to change. What do you think @onmyway133 @zenangst?

File: Sources/BarcodeScannerController.swift (hunk 1 above)
Comment (vadymmarkov) [reply Sources/BarcodeScannerController.swift]: I think it was here for the case when you have `UINavigationController` and push barcode scanner controller, so you don't end up with two "navigation bars". But arguably it's something you can manually control in the code.

File: BarcodeScanner.xcodeproj/project.pbxproj (hunk 2 above)
Comment (onmyway133) [reply BarcodeScanner.xcodeproj/project.pbxproj]: I don't have strong opinion about this. But as there's no reason to raise to 9.3, then 9.0 is fine I think

File: Sources/BarcodeScannerController.swift (hunk 1 above)
Comment (chadcummings) [reply Sources/BarcodeScannerController.swift]: I'll see if I can test this out by enhancing the example app to verify. I'll fix it if necessary.

File: BarcodeScanner.xcodeproj/project.pbxproj (hunk 2 above)
Comment (chadcummings) [reply BarcodeScanner.xcodeproj/project.pbxproj]: Sounds good. I'll move it to 9.0.

**Comment by Patchivic** [comment]
great @chadcummings 
I liked it :)

File: BarcodeScanner.xcodeproj/project.pbxproj (hunk 2 above)
Comment (zenangst) [reply BarcodeScanner.xcodeproj/project.pbxproj]: @vadymmarkov yeah we should totally change that. But if we don't need `9.3` changes in this PR then it should still use `9.0` :)

**Comment by zenangst** [comment]
I like this PR!

File: Sources/BarcodeScannerController.swift (hunk 1 above)
Comment (vadymmarkov) [reply Sources/BarcodeScannerController.swift]: @chadcummings Keep us posted. It's the only thing we'd like to verify before merging this PR.

**Comment by chadcummings** [comment]
Sorry, everyone. Work got a little busy for a bit.

I've changed the target to 9.0 and that all works well.

I enhanced the example project to demonstrate push and present navigation to ensure we didn't have a problem with "double navigation bars" with a line of code I removed.
It's working well with push navigation but for some reason the flashButton has now disappeared when pushing and presenting the BarcodeScannerController.

I'll see if I can get that fixed quick so we can finish up this PR.

**Comment by chadcummings** [comment]
Ooookkaaayyy, That's embarrassing. Apparently my iPad physically doesn't have a a light with the camera. Which explains why the flashButton doesn't show up. Problem solved.

**Comment by chadcummings** [comment]
Final fixes done. Let me know if there is anything else to do. Thanks.

**Comment by vadymmarkov** [comment]
@chadcummings Awesome! Thanks for your contribution.

**Comment by vadymmarkov** [comment]
Version 3.0.0 is out.
